### ISSUE TITLE
Read Query with Include can cause deadlock when reaching max DB sessions count.

### ISSUE THREAD

**Issue by alyaros** [dotnet/efcore#14759]
Version 2.1.1

Scenario I'm having:
1. I'm using EF Core to read from the DB
2. The read query has Include which cause it to actually generate 2 queries to the DB
3. The 1st is running and holding a DataReader (hence a db session) 
3. Next the 2nd generated query (triggered by Include) is running and trying to get a DB DataReader (session) while keeping the 1st open with session blocking , waiting for the 2nd query to complete.
4. In case I run many queries in parallel with Include - there are many sessions that are created.
5. Such that I reach the max sessions count of the DB and now the (2nd) Include query can't run because it has no available session and the 1st query is holding session waiting for the 2nd to complete. 

Question\Solution - Why the 1st query session\DataReader remains open while waiting to the 2nd query (Include) to complete.

MaximumSessions` = 1; 
var tasks = new List<Task>();
for (int i = 0; i < 2; i++)
 tasks.Add(InnerReadRecords(connectionString));

await Task.WhenAll(tasks);
Console.WriteLine("Reading Completed Successfully");

private static async Task InnerReadRecords(string connectionString)
{
            using (var ctx = new AppDatabaseContext(connectionString))
            {
                var randomRecords = await ctx.Blogs.Include(x => x.Posts).Take(100).ToListAsync();
                Console.WriteLine(randomRecords.Count);
            }
}`

**Comment by smitpatel** [comment]
Which database provider are you using?

**Comment by alyaros** [comment]
I'm using a custom Spanner driver.
But the close/dispose of readers are triggered by EF so the problem is
relevant to any provider.

**Comment by ajcvickers** [comment]
@alyaros EF Core does should not be leaving data readers open in this situation, and with other providers it does not appear to do so. Would it be possible to provide a small, runnable project/solution that reproduces the behavior you are seeing? Ideally if this is not a provider issue, then we should be able to reproduce with the SQL Server, SQLite, or in-memory providers.

**Comment by alyaros** [comment]
I'll try to give an example.

The following code get stuck in a deadlock.
Note: the .Include(..) part in the read query - Without it everything works correctly.

Why is it happening?
The 1st query runs and holds the reader open, next the 2nd query runs and try to get an open session \ reader, but it can't, this is because I've limited (in this scenario) the max session to 1 and set it "block" the thread when a session is requested and max has been reached. The question is why the 1st query keeps the reader open ? (dispose will be called only after the 2nd query will complete and than the reader \ session will be released)

In case I'll raise the max number of allowed sessions to 2 everything will work as expected.
both queries will complete, both will be disposed and I'll get the results as expected.

	Main()
	{
            SessionPoolManager.Default.SessionPoolOptions.MaximumActiveSessions = 1;
            SessionPoolManager.Default.SessionPoolOptions.WaitOnResourcesExhausted = ResourcesExhaustedBehavior.Block;

            try
            {
                var tasks = new List<Task>();
                for (int i = 0; i < 1; i++)
                    tasks.Add(InnerReadRecords(connectionString));

                await Task.WhenAll(tasks);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }            

            Console.WriteLine("Reading Completed Successfully");
        }

        private static async Task InnerReadRecords(string connectionString)
        {
            using (var ctx = new AppDatabaseContext(connectionString))
            {
                var randomRecords = await ctx.Blogs.Include(x => x.Posts).Take(100).ToListAsync();
                Console.WriteLine(randomRecords.Count);
            }
        }

**Comment by ajcvickers** [comment]
@alyaros It sounds like your provider doesn't really support multiple active result sets, but is probably indicating to EF Core that it does by returning `true` from `RelationalConnection.IsMultipleActiveResultSetsEnabled`.

**Comment by alyaros** [comment]
The provider I'm using does support multiple-sessions\ActiveResultSets but I've limited it to 1 session to demonstrate my point.

Do you agree that even If i'm limiting the underlying provider to 1 single session at maximum and I run a single query with Include phrase in it - the EF shouldn't get into a dead lock? 

Even if I set RelationalConnection.IsMultipleActiveResultSetsEnabled to True & my provider defined to support only 1 live session at a time.

It is happening because the Initial query holds the session open **UNTIL** the sub-query completes.
It can only happen if it was designed\or accidentally **WAIT** for the sub-query to complete.

Deadlock could not occur in any other case. If you run them in parallel or one after another or any order for that matter - if each query run and close it's session regardless the other - then a deadlock couldn't occur.

**Comment by smitpatel** [comment]
Duplicate of #12098

**Comment by smitpatel** [comment]
From my understanding, essentially you are trying to propose for single query rather than split query. There is no actionable bug here. Since EF Core is using multiple queries for certain queries and user is explicitly forcing to have only 1 query then it is going to deadlock. Which is true for all theoretical purposes. 

We are moving to single query in 3.0 timeframe anyway which will resolve this issue.

**Comment by ajcvickers** [comment]
@smitpatel In addition to Smit's comments, it seems to me like the multiple result sets capability as required by EF Core when `RelationalConnection.IsMultipleActiveResultSetsEnabled` is set is not something that your provider can support. Specifically, I mean that if it can sometimes deadlock when creating a new result set while there is already a result set in use on the current connection, then that's not sufficient for EF. If you get the same behavior with `RelationalConnection.IsMultipleActiveResultSetsEnabled` set to false, then we're really going to need some runnable code to reproduce it, but as Smit indicates, we don't intend to use this behavior in 3.0 anyway.
